lib/barcodevalidation/gtin.rb

```diff
@@ -1,46 +1,20 @@
 # frozen_string_literal: true
 
 require "forwardable"
+require_relative "invalid_gtin"
+require_relative "gtin/base"
+require_relative "gtin/check_digit"
+require_relative "gtin/gtin8"
+require_relative "gtin/gtin12"
+require_relative "gtin/gtin13"
+require_relative "gtin/gtin14"
 
 module BarcodeValidation
-  class GTIN < BarcodeValidation::DigitSequence
-    extend Forwardable
-    include Mixin::HasCheckDigit
-
-    VALID_LENGTHS = [8, 12, 13, 14].freeze
-
+  module GTIN
     def self.new(input)
-      super
-    rescue BarcodeValidation::Error => e
-      InvalidGTIN.new(input, error: e)
-    end
-
-    def valid?
-      VALID_LENGTHS.include?(length) && check_digit.valid?
-    end
-
-    class CheckDigit < DelegateClass(Digit)
-      include Adamantium::Flat
-      include Mixin::ValueObject
-
-      attr_reader :actual, :expected
-
-      def initialize(actual, expected: nil)
-        expected = actual if expected.nil?
-        @expected = Digit.new(expected)
-        @actual = Digit.new(actual)
-        super(@actual)
-      end
-
-      def valid?
-        actual == expected
-      end
-
-      def inspect
-        return super if valid?
-
-        "#<#{self.class}(#{actual}) invalid: expected #{expected}>"
-      end
+      module_eval("GTIN#{input.to_s.size}", __FILE__, __LINE__).new(input)
```

Comment:
i’d rather this was a boring case statement than used eval.

Reply:
Really? Not a fan of case switches, or even conditionals if I can at all avoid them, but OK.

Reply:
Heh, tried it, but rubocop's `Metrics/MethodLength` fails with "Method has too many lines. [12/10]" :) 

Bump it up here, or go through another PR for the rubocop settings?

Reply:
Actually, even though I'm not a fan, with 2584faa it would actually look OK: _(just long, and repetitive)_

```ruby
def self.new(input)
  case input.to_s.size
  when GTIN8::VALID_LENGTH
    GTIN8
  when GTIN12::VALID_LENGTH
    GTIN12
  when GTIN13::VALID_LENGTH
    GTIN13
  when GTIN14::VALID_LENGTH
    GTIN14
  else
    BarcodeValidation::InvalidGTIN
  end.new(input)
end
```

It's explicit without using magic numbers, but we'd need to adjust rubocop.

Reply:
Actually, pardon the comment spam, but rubocop is happy with a hash pattern:

```ruby
def self.new(input)
  klass = {
    8 => GTIN8,
    12 => GTIN12,
    13 => GTIN13,
    14 => GTIN14,
  }[input.to_s.size] || BarcodeValidation::InvalidGTIN

  klass.new(input)
end
```

That actually looks neater to me than the case switch, what do you think?

Reply:
make the hash a constant and it’s pretty much equivalent in c.

Reply:
(i’d tend to use `when then` to avoid the long method but this also works)

Reply:
and if possible i tend to prefer not using branches but eval is not something id use in favour of branches

Reply:
OK, hoping what I've just pushed up is a happy medium:

```ruby
def new(input)
  (class_for_input(input) || BarcodeValidation::InvalidGTIN).new(input)
end

private

def class_for_input(input)
  [GTIN8, GTIN12, GTIN13, GTIN14].find do |klass|
    input.to_s.size == klass::VALID_LENGTH
  end
end
```

Not as concise as eval, but also not as verbose as a case switch.